File: tests/07_phases_ompss.dg/fortran/success_task_03.f90
! <testinfo>
! test_generator="config/mercurium-ompss no-nanos6"
! </testinfo>

SUBROUTINE F(X, JSIZE, ISIZE)
    INTEGER, INTENT(IN) :: JSIZE, ISIZE
    INTEGER, INTENT(INOUT) :: X(JSIZE, ISIZE)
    X = 0
END SUBROUTINE F

SUBROUTINE G(X, J, I, N)
    INTEGER, INTENT(IN) :: N, I, J
    INTEGER, INTENT(INOUT) :: X(N, N)
    INTEGER :: JSIZE, ISIZE

    INTERFACE
        SUBROUTINE F(X, JSIZE, ISIZE)
            INTEGER, INTENT(IN) :: JSIZE, ISIZE
            INTEGER, INTENT(INOUT) :: X(JSIZE, ISIZE)
        END SUBROUTINE F
    END INTERFACE

    JSIZE = N - J + 1
    ISIZE = N - I + 1
    !$OMP TASK INOUT(X(J:, I:))
    CALL F( X(J, I),  JSIZE, ISIZE)
    !$OMP END TASK
END SUBROUTINE G


PROGRAM P
    IMPLICIT NONE
    INTEGER, PARAMETER :: N = 5
    INTEGER :: X(N, N)
    INTEGER :: I, J
    INTEGER :: IND_I, IND_J


    IND_I = 3
    IND_J = 1

    DO  I = 1, N
        DO  J = 1, N
            X(J, I) = ((I-1)*N + J)
        ENDDO
    ENDDO

    CALL G(X, IND_J, IND_I, N)
    !$OMP TASKWAIT

    DO I=1, IND_I - 1
        DO J = 1, IND_J - 1
            IF (X(J, I) /= ((I-1)*N + J)) STOP 1
        ENDDO
    ENDDO
    IF (ANY(X(IND_J:, IND_I:) /= 0)) STOP 2
END PROGRAM P

File: tests/07_phases_ompss.dg/c/success_task_03.c
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/



/*
<testinfo>
test_generator=config/mercurium-ompss
</testinfo>
*/
#include <stdlib.h>

int main(int arg, char* argv[])
{
    int x[10][20];

    int i, j;
    for (i = 0; i < 10; i++)
    {
        for (j = 0; j < 20; j++)
        {
            x[i][j] = 1;
        }
    }

#pragma omp task inout(x[1][2])
    {
        x[1][2] += 41;
    }

#pragma omp task in(x[1][2])
    {
        if (x[1][2] != 42) abort();
    }

#pragma omp taskwait
    return 0;
}
